aws-s3-transfer-manager/src/io/stream.rs

```diff
@@ -94,13 +133,110 @@ pub(super) enum RawInputStream {
     Buf(Bytes),
     /// File based input
     Fs(PathBody),
+    /// User provided custom stream
+    Dyn(BoxStream),
+}
+
+/// The context of an input stream.
+#[derive(Debug)]
+pub struct StreamContext {
+    part_size: usize,
+}
+
+impl StreamContext {
+    pub(super) fn new(part_size: usize) -> Self {
+        Self { part_size }
+    }
+
+    /// The part size to use when yielding parts.
+    /// NOTE: this _may_ differ from the configured part size (e.g. if the target part size would
+    /// result in exceeding the maximum number of parts allowed).
+    pub fn part_size(&self) -> usize {
+        self.part_size
+    }
+
+    // TODO - eventually make the ability to allocate a buffer public after carefully review of the `Buffer` API.
+    /// Request a new buffer to fill
+    pub(crate) fn new_buffer(&self, capacity: usize) -> Buffer {
+        // TODO - replace allocation with memory pool
+        Buffer::new(capacity)
+    }
+}
+
+/// Contents and (optional) metadata for a single part of a [multipart upload].
+///
+/// [multipart upload]: https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PartData {
+    // 1-indexed
+    pub(crate) part_number: u64,
+    pub(crate) data: Bytes,
+}
+
+impl PartData {
+    /// Create a new part
+    pub(super) fn new(part_number: u64, data: impl Into<Bytes>) -> Self {
+        Self {
+            part_number,
+            data: data.into(),
+        }
+    }
+}
+
+/// Trait representing a stream of object parts (streaming body).
+///
+/// Individual parts are streamed via the `poll_part` function, which asynchronously yields
+/// instances of `PartData`. When `Poll::Ready(None)` is returned the stream is assumed to have
+/// reached EOF and is finished.
+///
+/// The `size_hint` function provides insight into the total number of bytes that will be streamed.
+pub trait PartStream {
+    /// Attempt to pull the next part from the stream.
+    ///
+    /// The `stream_cx` will have the part size that should be utilized. Implementations should be
+    /// careful to only yield full parts for every part except the last one, which _may_ be less
+    /// than the full part size.
+    fn poll_part(
+        self: std::pin::Pin<&mut Self>,
```

Comment:
since this is mutable, we can't do concurrent `poll_read()` calls, right?

So if, in the future, we wanted to let users build their own custom type that can do concurrent reads, we'd need to create a new trait for that, right?

It's not something we could add to this trait by introducing some extra fn like `allows_concurrency(&self) -> bool`?

Just bringing up more future design concerns, and trying to understand how they might be tackled

Reply:
> since this is mutable, we can't do concurrent poll_read() calls, right?

Correct. Most `Poll*` methods (e.g. [Future](https://doc.rust-lang.org/std/future/trait.Future.html), [tokio::AsyncRead](https://docs.rs/tokio/latest/tokio/io/trait.AsyncRead.html), etc usually take a `Pin<&mut Self>`). This is why we have to wrap a `tokio::sync::Mutex` around implementations of this though to provide that. Most streams are sequential and can't be read concurrently anyway (at least not without some kind of synchronization). 

> It's not something we could add to this trait by introducing some extra fn like allows_concurrency(&self) -> bool?

No, not like that but I something like the channel based stream I mentioned might work and allow people to read data however they want.

Reply:
Downside of channel-based-stream is that it's up to the user to pick the ideal number of parts-in-memory, since they're creating the channel, and spawning the tasks that feed it before the poll_read() is actually called.

If we had a different trait, and knew the user could provide parts in parallel, the transfer manager, with its global view of work, would have more control over the parts-in-memory.

we can tackle this later, just airing thoughts